static/app/utils/profiling/profile/jsSelfProfile.tsx

```diff
@@ -0,0 +1,153 @@
+import {lastOfArray} from 'sentry/utils';
+import {CallTreeNode} from 'sentry/utils/profiling/callTreeNode';
+import {Frame} from 'sentry/utils/profiling/frame';
+
+import {stackMarkerToHumanReadable} from './../formatters/stackMarkerToHumanReadable';
+import {resolveJSSelfProfilingStack} from './../jsSelfProfiling';
+import {Profile} from './profile';
+import {createFrameIndex} from './utils';
+
+export class JSSelfProfile extends Profile {
+  static FromProfile(profile: JSSelfProfiling.Trace): JSSelfProfile {
+    const frameIndex = createFrameIndex(profile.frames, profile);
+    // In the case of JSSelfProfiling, we need to index the abstract marker frames
+    // as they will otherwise not be present in the ProfilerStack.
+    const markers: JSSelfProfiling.Marker[] = [
+      'gc',
+      'layout',
+      'other',
+      'paint',
+      'script',
+      'style',
+    ];
+
+    for (const marker of markers) {
+      frameIndex[marker] = new Frame({
+        key: marker,
+        name: stackMarkerToHumanReadable(marker),
+        line: undefined,
+        column: undefined,
+        is_application: false,
+      });
+    }
+
+    const startedAt = profile.samples[0].timestamp;
+    const endedAt = lastOfArray(profile.samples).timestamp;
+
+    const jsSelfProfile = new JSSelfProfile(
+      endedAt - startedAt,
+      startedAt,
+      endedAt,
+      'JSSelfProfiling',
+      'milliseconds'
+    );
+
+    // Because JS self profiling takes an initial sample when we call new Profiler(),
+    // it means that the first sample weight will always be zero. We want to append the sample with 0 weight,
+    //  because the 2nd sample may part of the first sample's stack. This way we keep the most information we can of the stack trace
+    jsSelfProfile.appendSample(
+      resolveJSSelfProfilingStack(
+        profile,
+        profile.samples[0].stackId,
+        frameIndex,
+        profile.samples[0].marker
+      ),
+      0
+    );
+
+    // We start at stack 1, because we've already appended stack 0 above. The weight of each sample is the
+    // difference between the current sample and the previous one.
+    for (let i = 1; i < profile.samples.length; i++) {
+      jsSelfProfile.appendSample(
+        resolveJSSelfProfilingStack(
+          profile,
+          profile.samples[i].stackId,
+          frameIndex,
+          profile.samples[i].marker
+        ),
+        profile.samples[i].timestamp - profile.samples[i - 1].timestamp
+      );
+    }
+
+    return jsSelfProfile.build();
+  }
+
+  appendSample(stack: Frame[], weight: number): void {
+    let node = this.appendOrderTree;
```

Comment:
The `appendOrderTree` uses `Frame.Root` initialized [here](https://github.com/getsentry/sentry/blob/2a7a0aac94e3b612cb49ca24c8882b7290c788f8/static/app/utils/profiling/profile/profile.tsx#L23) which has `Root.type` set to `mobile` by default. Is this important here?

Reply:
@Zylphrex correct, the root is just there as a meta node and shouldn't get rendered anywhere :) Thanks for reviewing!